internal/driver/onvifdiscovery_test.go

```diff
@@ -0,0 +1,349 @@
+			devices:           createTestDeviceList(),
+			discoveredDevices: createDiscoveredList(),
+			filtered:          []sdkModel.DiscoveredDevice{},
+			nameCalls:         4,
+		},
+		{
+			name: "All new devices",
+			devices: []contract.Device{
+				{
+					Name: "device-onvif-camera", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "abc",
+						},
+					},
+				},
+			},
+			discoveredDevices: createDiscoveredList(),
+			filtered:          createDiscoveredList(),
+			nameCalls:         1,
+		},
+		{
+			name:    "new and old devices",
+			devices: createTestDeviceList(),
+			discoveredDevices: []sdkModel.DiscoveredDevice{
+				{
+					Name: "testDevice1", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "123",
+						},
+					},
+				},
+				{
+					Name: "testDevice2", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "456",
+						},
+					},
+				},
+				{
+					Name: "testDevice3", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "789",
+						},
+					},
+				},
+				{
+					Name: "testDevice4", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "xyz",
+						},
+					},
+				},
+				{
+					Name: "testDevice5", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "def",
+						},
+					},
+				},
+			},
+			filtered: []sdkModel.DiscoveredDevice{
+				{
+					Name: "testDevice4", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "xyz",
+						},
+					},
+				},
+				{
+					Name: "testDevice5", Protocols: map[string]models.ProtocolProperties{
+						OnvifProtocol: map[string]string{
+							EndpointRefAddress: "def",
+						},
+					},
+				},
+			},
+			nameCalls: 4,
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			driver, mockService := createDriverWithMockService()
+			mockService.On("Devices").
+				Return(test.devices).Times(2)
+			mockService.On("Name").
+				Return("device-onvif-camera").Times(2 * test.nameCalls)
+			filtered := driver.discoverFilter(test.discoveredDevices)
+			mockService.AssertExpectations(t)
+
+			assert.Equal(t, test.filtered, filtered)
+		})
+	}
+}
+
+// func createParams() (params onvif.DeviceParams) {
+// 	return onvif.DeviceParams{
+// 		Xaddr:              "1.1.1.1:1",
+// 		EndpointRefAddress: "1234",
+// 		Username:           "hello",
+// 		Password:           "world",
+// 	}
+// }
+
+// func createDevice() onvif.Device {
+// 	device, _ := onvif.NewDevice(createParams())
+// 	return *device
+// }
+
+// func TestmockService_createDiscoveredDevice(t *testing.T) {
```

Comment:
i think this fails because the code is trying to communicate with an actual camera at that xaddr

Reply:
Yes. That is something that I started but will get back to now that I have some more knowledge on mocking